lib/Decl.cpp

```diff
@@ -188,4 +195,135 @@ llvm::Value *FunctionDecl::CallFromLiftedBlock(
+
+    json.insert(llvm::json::Object::KV{
+        llvm::json::ObjectKey("return stack pointer"),
+        llvm::json::Value(std::move(return_stack_pointer_json))});
+  }
+  return json;
+}
+
+// Serialize a ParameterDecl to JSON
+llvm::json::Object ParameterDecl::SerializeToJSON(const llvm::DataLayout &dl) {
+  // Get the serialization for the ValueDecl
+  ValueDecl *val_decl_ptr = dynamic_cast<ValueDecl *>(this);
+  llvm::json::Object param_json = val_decl_ptr->SerializeToJSON(dl);
+
+  // Insert "name"
+  param_json.insert(
+      llvm::json::Object::KV{llvm::json::ObjectKey("name"), this->name});
+
+  return param_json;
+}
+
+// Serialize a ValueDecl to JSON
+llvm::json::Object ValueDecl::SerializeToJSON(const llvm::DataLayout &dl) {
+  llvm::json::Object value_json;
+
+  if (this->reg) {
+    // The value is in a register
+    value_json.insert(llvm::json::Object::KV{llvm::json::ObjectKey("register"),
+                                             this->reg->name});
+  } else if (this->mem_reg) {
+    // The value is in memory
+    llvm::json::Object memory_json;
+    memory_json.insert(llvm::json::Object::KV{llvm::json::ObjectKey("register"),
+                                              this->mem_reg->name});
+    memory_json.insert(llvm::json::Object::KV{llvm::json::ObjectKey("offset"),
+                                              this->mem_offset});
+
+    // Wrap the memory_json structure in a memory block
+    value_json.insert(
+        llvm::json::Object::KV{llvm::json::ObjectKey("memory"),
+                               llvm::json::Value(std::move(memory_json))});
+  } else {
+    LOG(ERROR) << "Trying to serialize a value that has not been allocated";
+    exit(1);
+  }
+
+  value_json.insert(llvm::json::Object::KV{llvm::json::ObjectKey("type"),
+                                           TranslateType(*this->type, dl)});
+
+  return value_json;
+}
+
+// Create a Function Declaration from llvm::Function
+FunctionDecl FunctionDecl::Create(const llvm::Function &func, const llvm::Module &mdl) {
+  const llvm::DataLayout dl = mdl.getDataLayout();
+  const remill::Arch* arch = remill::Arch::GetModuleArch(mdl);
+
+  FunctionDecl decl;
+  decl.type = func.getFunctionType();
+  
+  decl.name = func.getName().data();
+  decl.demangled_name = llvm::demangle(func.getName().data());
+  decl.dl = &dl;
+
+  // Try to guess the parameter and return value register allocation by looking
+  // at the calling convention of the function
+  llvm::CallingConv::ID cc_id = func.getCallingConv();
+  std::unique_ptr<CallingConvention> cc = nullptr;
+
+  switch (cc_id) {
+    case llvm::CallingConv::X86_64_SysV:
+      cc = std::unique_ptr<anvill::X86_64_SysV>(new X86_64_SysV(arch));
+    default:
+      // TODO(aty): find a better way to do this since the calling conventions given
+      // by llvm::function cannot be trusted
+      cc = std::unique_ptr<anvill::X86_C>(new X86_C(arch));
```

Comment:
I would check the arch name / OS to decide on the most suitable fallback ABI here. It doesn't make sense to use a 32-bit ABI if we're on a 64-bit arch, for example.